torch/utils/data/dataloader.py

```diff
@@ -242,11 +286,25 @@ def _shutdown_workers(self):
         if not self.shutdown:
             self.shutdown = True
             self.done_event.set()
+            # if worker_manager_thread is waiting to put
+            if not self.data_queue.empty():
+                self.data_queue.get()
             for _ in self.workers:
                 self.index_queue.put(None)
+            # if all workers hang, no None is sent to worker_manager_thread, we
+            # put None to let worker_manager_thread exit
+            # empty check prevents put from hanging
+            if self.worker_result_queue.empty():
```

Comment:
I think I mentioned in one of the previous PRs that having non-empty queue doesn't imply that there's None in it. I think it's ok to just add it unconditionally (maybe check that it's not full to avoid blocking)

Reply:
Yes, I remember the comment. I didn't remove it exactly because of worrying about blocking. I'll see if I can check if it is full.